test/e2e/app-dir/app-root-params/use-cache.test.ts

```diff
@@ -0,0 +1,79 @@
+import { nextTestSetup } from 'e2e-utils'
+import { join } from 'path'
+import { createSandbox } from 'development-sandbox'
+
+describe('app-root-params - cache - at runtime', () => {
+  const { next, isNextDev } = nextTestSetup({
+    files: join(__dirname, 'fixtures', 'use-cache-runtime'),
+    skipStart: true,
+  })
+
+  if (isNextDev) {
+    it('should error when using rootParams within a "use cache" - dev', async () => {
+      await using sandbox = await createSandbox(
+        next,
+        undefined,
+        '/en/us/use-cache'
+      )
+      const { session } = sandbox
+      await session.assertHasRedbox()
+      expect(await session.getRedboxDescription()).toInclude(
+        'Route /[lang]/[locale]/use-cache used "unstable_rootParams" inside `"use cache"` or `unstable_cache`'
+      )
+    })
+
+    it('should error when using rootParams within `unstable_cache` - dev', async () => {
+      await using sandbox = await createSandbox(
+        next,
+        undefined,
+        '/en/us/unstable_cache'
+      )
+      const { session } = sandbox
+      await session.assertHasRedbox()
+      expect(await session.getRedboxDescription()).toInclude(
+        'Route /[lang]/[locale]/unstable_cache used "unstable_rootParams" inside `"use cache"` or `unstable_cache`'
+      )
+    })
+  } else {
+    beforeAll(async () => {
+      await next.start()
+    })
+
+    it('should error when using rootParams within a "use cache" - start', async () => {
+      await next.render$('/en/us/use-cache')
+      expect(next.cliOutput).toInclude(
+        'Error: Route /[lang]/[locale]/use-cache used "unstable_rootParams" inside `"use cache"` or `unstable_cache`.'
+      )
+    })
+
+    it('should error when using rootParams within `unstable_cache` - start', async () => {
+      await next.render$('/en/us/unstable_cache')
+      expect(next.cliOutput).toInclude(
+        'Error: Route /[lang]/[locale]/unstable_cache used "unstable_rootParams" inside `"use cache"` or `unstable_cache`.'
+      )
+    })
+  }
+})
+
+describe('app-root-params - cache - at build', () => {
```

Comment:
can potentially move to `test/production/app-dir` where only gets run with production tests